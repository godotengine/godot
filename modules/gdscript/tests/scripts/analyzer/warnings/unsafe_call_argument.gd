func variant_func(x: Variant) -> void:
	print(x)

func int_func(x: int) -> void:
	print(x)

func float_func(x: float) -> void:
	print(x)

func node_func(x: Node) -> void:
	print(x)

# We don't want to execute it because of errors, just analyze.
func no_exec_test():
	var variant: Variant = null
	var untyped_int = 42
	var untyped_string = "abc"
	var variant_int: Variant = 42
	var variant_string: Variant = "abc"
	var typed_int: int = 42

	variant_func(untyped_int) # No warning.
	variant_func(untyped_string) # No warning.
	variant_func(variant_int) # No warning.
	variant_func(variant_string) # No warning.
	variant_func(typed_int) # No warning.

	int_func(untyped_int)
	int_func(untyped_string)
	int_func(variant_int)
	int_func(variant_string)
	int_func(typed_int) # No warning.

	float_func(untyped_int)
	float_func(untyped_string)
	float_func(variant_int)
	float_func(variant_string)
	float_func(typed_int) # No warning.

	node_func(variant)
	node_func(Object.new())
	node_func(Node.new()) # No warning.
	node_func(Node2D.new()) # No warning.

	# GH-82529
	print(Callable(self, "test")) # No warning.
	print(Callable(variant, "test"))

	print(Dictionary(variant))
	print(Vector2(variant))
	print(int(variant))

func test():
	pass
