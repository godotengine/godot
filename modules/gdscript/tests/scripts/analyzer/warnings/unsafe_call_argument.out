GDTEST_OK
>> WARNING
>> Line: 28
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "int_func()" requires the subtype "int" but the supertype "Variant" was provided.
>> WARNING
>> Line: 29
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "int_func()" requires the subtype "int" but the supertype "Variant" was provided.
>> WARNING
>> Line: 30
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "int_func()" requires the subtype "int" but the supertype "Variant" was provided.
>> WARNING
>> Line: 31
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "int_func()" requires the subtype "int" but the supertype "Variant" was provided.
>> WARNING
>> Line: 34
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "float_func()" requires the subtype "float" but the supertype "Variant" was provided.
>> WARNING
>> Line: 35
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "float_func()" requires the subtype "float" but the supertype "Variant" was provided.
>> WARNING
>> Line: 36
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "float_func()" requires the subtype "float" but the supertype "Variant" was provided.
>> WARNING
>> Line: 37
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "float_func()" requires the subtype "float" but the supertype "Variant" was provided.
>> WARNING
>> Line: 40
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "node_func()" requires the subtype "Node" but the supertype "Variant" was provided.
>> WARNING
>> Line: 41
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the function "node_func()" requires the subtype "Node" but the supertype "Object" was provided.
>> WARNING
>> Line: 47
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the constructor "Callable()" requires the subtype "Object" but the supertype "Variant" was provided.
>> WARNING
>> Line: 49
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the constructor "Dictionary()" requires the subtype "Dictionary" but the supertype "Variant" was provided.
>> WARNING
>> Line: 50
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the constructor "Vector2()" requires the subtype "Vector2" or "Vector2i" but the supertype "Variant" was provided.
>> WARNING
>> Line: 51
>> UNSAFE_CALL_ARGUMENT
>> The argument 1 of the constructor "int()" requires the subtype "int", "bool", or "float" but the supertype "Variant" was provided.
